infra/terraform/modules/kubernetes_cluster/bin/kops-specs/generate_specs.py

```diff
@@ -0,0 +1,211 @@
+  return dumper.represent_scalar('tag:yaml.org,2002:str', data)
+
+yaml.add_representer(str, str_presenter)
+
+
+parser = argparse.ArgumentParser()
+
+parser.add_argument('--template-path', nargs='?', required=True)
+parser.add_argument('--out-path', nargs='?', required=True)
+parser.add_argument('--cluster-name', nargs='?', required=True)
+parser.add_argument('--vpc-id', nargs='?', required=True)
+parser.add_argument('--state-bucket', nargs='?', required=True)
+parser.add_argument('--kubernetes-version', nargs='?', required=True)
+parser.add_argument('--dns-zone', nargs='?', required=True)
+parser.add_argument('--zones', nargs='?', required=True)
+parser.add_argument('--network-cidr', nargs='?', required=True)
+parser.add_argument('--node-instance-type', nargs='?', required=True)
+parser.add_argument('--node-count', type=int, nargs='?', required=True)
+parser.add_argument('--node-volume-size', type=int, nargs='?', required=True)
+parser.add_argument('--ami-name', nargs='?', required=True)
+parser.add_argument('--bastion-instance-type', nargs='?', required=True)
+parser.add_argument('--bastion-count', type=int, nargs='?', required=True)
+parser.add_argument('--master-instance-type', nargs='?', required=True)
+parser.add_argument('--public-subnet-zones', type=json.loads, nargs='?',
+                    required=True)
+parser.add_argument('--private-subnet-zones', type=json.loads, nargs='?',
+                    required=True)
+parser.add_argument('--public-subnet-cidrs', type=json.loads, nargs='?',
+                    required=True)
+parser.add_argument('--private-subnet-cidrs', type=json.loads, nargs='?',
+                    required=True)
+parser.add_argument('--nat-gateway-subnets', type=json.loads, nargs='?',
+                    required=True)
+
+
+args = parser.parse_args()
+
+if not os.path.exists(args.out_path):
+    os.makedirs(args.out_path)
+
+def template(fname):
+    return os.path.join(args.template_path, fname)
+
+def outfile(fname):
+    return os.path.join(args.out_path, fname)
+
+
+zones = sorted(args.zones.split(','))
+
+nat_gateway_zones = {
+    args.public_subnet_zones[k] : v
+    for k, v in args.nat_gateway_subnets.items()
+}
+
+subnets = {
+    'public': [],
+    'private': []
+}
+for kind in ['private', 'public']:
+    cidrs = vars(args)["{}_subnet_cidrs".format(kind)]
+    subnet_zones = vars(args)["{}_subnet_zones".format(kind)]
+
+    for subnet_id, zone in sorted(subnet_zones.items(), key=lambda i: i[1]):
+        subnet = {
+            'id': subnet_id,
+            'zone': zone,
+            'cidr': cidrs[subnet_id],
+            'type': 'Private' if kind == 'private' else 'Utility',
+            'name': zone if kind == 'private' else 'utility-{}'.format(zone)
+        }
+
+        if kind == 'private':
+            subnet['egress'] = nat_gateway_zones[zone]
+
+        subnets[kind].append(subnet)
+
+# Cluster spec
+with open(template('cluster.tpl.yml'), 'r') as stream:
```

Comment:
Is this related to your comment regarding template files not used anymore?

I see we read some YAML file and changes some of it before writing to the output YAML. Is the template YAML used anywhere else a part from this script?

(Sorry, still going through this mammoth PR)

Reply:
Are we keeping the YAML files so that we can change parts of them without requiring changes to this script? I wonder if this may be confusing at some point (e.g. we update the template YAML but the script still override it)

(don't change anything, it's more an observation than anything)

Reply:
Yes, there's now this python script generating kops YAML files, instead of Go templates rendering them. This script uses the files in `infra/terraform/modules/kubernetes_cluster/data/kops` as a starting point, and populates them with output values from other terraform-managed AWS resources (subnet IDs, security group IDs, etc) - so conceptually it's similar to the Go template version, but doesn't have the limitations imposed by a template language.

The idea is that the `*.tpl.yml` files will contain options that aren't environment-specific, like flags on the k8s API server, so they remain consistent across environments. That does mean that template files like `bastions.tpl.yml` contain pretty much nothing, but it seems best to use the same approach for all kops resources rather than have some entirely generated in python.

Reply:
The python script only ever inserts values that are either variables of the `kubernetes_cluster` TF module, or IDs of AWS resources. The YAML files are always rendered with this script on every `terraform apply`, so changes to either the template YAML or variables/AWS resources are applied together. That said, this module does need some detailed documentation to make it clear how this all works.